File: Leetcode/206/206.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	if head.Next == nil {
		return head
	} else if head.Next.Next == nil {
		temp := head.Next
		head.Next = temp.Next
		temp.Next = head
		return temp
	}
	node := reverseList(head.Next)
	node.Next = head
	head.Next = nil
	return node
}

File: Leetcode/206/206.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList2(struct ListNode* head){
    if (head == NULL) {
        return NULL;
    }
    if (head->next == NULL) {
        return head;
    }
    struct ListNode* ptr = reverseList2(head->next);
    head->next->next = head;
    head->next = NULL;
    return ptr;
}

struct ListNode* reverseList(struct ListNode* head){
    if (head == NULL) {
        return NULL;
    }
    if (head->next == NULL) {
        return head;
    }
    struct ListNode* prev = NULL;
    struct ListNode* curr = head;
    while (curr != NULL) {
        struct ListNode* temp = curr->next;
        curr->next = prev;
        prev = curr;
        curr = temp;
    }
    return prev;
}